### PR TITLE
Fix insertion of None value in MultiIndex

### LINKED ISSUE #59003: BUG: None value in MultiIndex can't be found
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

indices = [["one"], ["a"], ["yes"]]
df = pd.DataFrame([0], index=indices).T
df["one", None, "yes"] = 1
print(df)
print()
print(df["one"])
print()
print(df["one"][None])
```

### Issue Description

The dataframe has an item at the multiindex of ["one", None, "yes"], but attempting to access it fails. It appears to only happen when I use None/nan values for the middle index that I'm adding, and it only happens if I add the value using the df[] = syntax. 

### Expected Behavior

It should return a value when I query df["one"][None]. If I construct the dataframe with this data in it to begin with, or add it with a df.concat, it works.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.9.2.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 4.18.0-513.24.1.el8_9.x86_64
Version               : #1 SMP Thu Mar 14 14:20:09 EDT 2024
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.2
numpy                 : 1.23.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 70.0.0
pip                   : 23.1
Cython                : None
pytest                : 7.4.4
hypothesis            : None
sphinx                : 7.3.7
blosc                 : None
feather               : None
xlsxwriter            : 3.2.0
lxml.etree            : 5.2.2
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.4
IPython               : 8.18.1
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : 2024.6.0
gcsfs                 : None
matplotlib            : 3.9.0
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : 3.1.2
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : 2.0.12
tables                : None
tabulate              : None
xarray                : None
xlrd                  : 2.0.1
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by chaoyihu**
- [x] closes #59003 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v3.0.0.rst [hunk 1]
```diff
@@ -543,6 +543,7 @@ MultiIndex
 ^^^^^^^^^^
 - :func:`DataFrame.loc` with ``axis=0``  and :class:`MultiIndex` when setting a value adds extra columns (:issue:`58116`)
 - :meth:`DataFrame.melt` would not accept multiple names in ``var_name`` when the columns were a :class:`MultiIndex` (:issue:`58033`)
+- :meth:`MultiIndex.insert` would not insert None value correctly at unified location of index -1 (:issue:`59003`)
```
Comment (yuanx749) [inline doc/source/whatsnew/v3.0.0.rst:546]:
```suggestion
- :meth:`MultiIndex.insert` would not insert ``None`` value correctly at unified location of index -1 (:issue:`59003`)
```
nitpick

**Review by yuanx749 — COMMENTED** [review]
Is it possible to also include other NaN in this fix? I think inserting ``np.nan`` would cause this bug too.

File: pandas/tests/test_multilevel.py [hunk 2]
```diff
@@ -288,6 +289,17 @@ def test_multiindex_with_na(self):
 
         tm.assert_frame_equal(df, expected)
 
+    def test_multiindex_insert_level_with_na(self):
+        # GH 59003
+        indices = [["one"], ["a"], ["yes"]]
+        df = DataFrame([0], index=indices).T
```
Comment (mroeschke) [inline pandas/tests/test_multilevel.py:295]: Could you construct this without `.T`?

File: pandas/tests/test_multilevel.py [hunk 3]
```diff
@@ -288,6 +289,17 @@ def test_multiindex_with_na(self):
...
+        df["one", None, "yes"] = 1
+        expected = DataFrame([1], index=Index(["yes"])).T
+        tm.assert_frame_equal(df["one"][None], expected)
```
Comment (mroeschke) [inline pandas/tests/test_multilevel.py:298]: Could you test this without doing a chained indexing?

File: pandas/tests/test_multilevel.py [hunk 4]
```diff
@@ -288,6 +289,17 @@ def test_multiindex_with_na(self):
...
+        df["one", np.nan, "yes"] = 1
```
Comment (mroeschke) [inline pandas/tests/test_multilevel.py:299]: Can you use `pytest.mark.parametrize` with `np.nan` and `None`

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @chaoyihu

**Pull request merged**